### ISSUE TITLE
Train a model

### ISSUE THREAD

**Issue by nsanjay01** [nsanjay01/B-U585I-IOT02A_tinyML_IMU#2]
Once you verify that our sensors are producing right 'live' data, then move onto training a model.

Anomaly detection -
1. Data Normalization: The pre-processing step
Raw output of sensor is in milli G, therefore convert it to small float value that is between -2.0 and 2.0
It can be done by dividing by 1000. This is known as scaling the data.
**Windowing** : model works best on series of data, not just a single data point.
So collect a window of 50 samples (50 x X,Y,Z ) to feed the model.

2. Which Model to use?
The autoencoder (neural network)
* How it works: we train the model to compress this input data and then recontruct it.
* Detection logic: when "Anomaly" data comes in, the model compares it with predicted pattern and we get recontruction error high.

3. How to train the model
We can use Google Colab with python.

The training pipeline:
1. Data collection: we can public datasets to make our model more robust. By combining our own data with public data, we prevent the model from "overfitting".

For anomaly detection using an IMU (accelerometer/Gyroscope), here are the public dataset

A. UCI Human Activity Recognition (HAR)

This is the "Hello World" of IMU datasets. It contains data from smartphones (accelerometer and gyro) for activities like walking, sitting, and standing.

    Why use it: It’s great for teaching your model what "general movement" looks like so it doesn't trigger a false anomaly every time you pick up the board.

    Link: [UCI Machine Learning Repository](https://archive.ics.uci.edu/ml/datasets/human+activity+recognition+using+smartphones)

B. Case Western Reserve University (CWRU) Bearing DatasetZ:

If your anomaly detection is for Industrial Maintenance (detecting if a motor or fan is failing via vibration):

    Why use it: It is the industry standard for vibration-based anomaly detection. It contains "Normal" motor data and "Faulty" data (ball innards, outer race faults).

    Link: [CWRU Bearing Data Center](https://engineering.case.edu/bearingdatacenter)

C. WISDM (Wireless Sensor Data Mining)

Similar to UCI HAR but includes more varied activities like jogging and climbing stairs.

    Link: [WISDM Project](https://www.cis.fordham.edu/wisdm/dataset.php)

2. How to Integrate Public Data with Your Own
You cannot just mix the files and hope for the best. You need a strategy called Data Alignment.

   1. **Coordinate Alignment**: Ensure the axes match. Your STM32 board might define "Z" as up, but a public dataset might define "Y" as up. You must swap the columns in your CSV to match.

  2.  **Sampling Rate Matching**: Your code currently sets the IMU to 104 Hz. If the public dataset was recorded at 50 Hz, you must "downsample" your data or "interpolate" theirs so the model sees the same "speed" of time.

 3. **Unit Conversion:** Most public data is in G's (9.81 m/s2). Your BSP returns mg (milli-G's).

        Calculation: Divide your raw BSP values by 1000 to match public data.

3. The "Transfer Learning" Strategy

Instead of mixing the data into one big pile, use this robust training method:

1.Pre-train on Public Data: Train your Autoencoder or Classifier on the large public dataset first. This teaches the model the "physics" of movement.

2. Fine-tune on Your Data: "Freeze" the first few layers of your model and train the last layer only on data from your specific B-U585I board. This specializes the model to your exact hardware.

4. Recommended Tools for Integration

To make this easy without writing 1,000 lines of Python:

 1. Edge Impulse (Data Sources): They have a "Data Sources" feature where you can upload public S3 buckets or OpenMV datasets directly alongside your uploaded STM32 data.

  2. TensorFlow Data Validation (TFDV): Use this Python tool to compare your data to public data. It will tell you if your data is "drifting" (e.g., if your sensor is much noisier than the public one).

Summary Checklist for Robustness
Action	           Purpose
Add Noise	Artificially add slight "jitter" to the public data to simulate your specific sensor noise.
Augmentation	Rotate the public data. If a person was walking North, flip the axes to simulate them walking South.
Normalization	         Use Min-Max Scaling so all data (yours and theirs) stays between -1 and 1.

**Comment by nsanjay01** [comment]
Using both sensors transforms your project from 3D tracking to 6D tracking. This provides the "full picture" of movement: the accelerometer tells the model about linear impact and gravity, while the gyroscope explains the rotation and "swing."

Here is how you need to update both your STM32 firmware and your Python training pipeline to handle this 6-axis data.

1. STM32 Code: Initializing & Logging 6-Axis Data

You need to enable the gyroscope and update your UART logging format so the Python script receives all 6 values (Acc X, Y, Z and Gyro X, Y, Z).
Update your MEMS_Init function:
```
static void MEMS_Init(void)
{
  // ... (previous io_ctx code) ...
  
  /* 1. Initialize and Enable Accelerometer at 52Hz */
  ISM330DHCX_ACC_SetOutputDataRate(&MotionSensor, 52.0f);
  ISM330DHCX_ACC_SetFullScale(&MotionSensor, 4); // +/- 4g
  ISM330DHCX_ACC_Enable(&MotionSensor);

  /* 2. Initialize and Enable Gyroscope at 52Hz */
  ISM330DHCX_GYRO_SetOutputDataRate(&MotionSensor, 52.0f);
  ISM330DHCX_GYRO_SetFullScale(&MotionSensor, 2000); // +/- 2000 dps
  ISM330DHCX_GYRO_Enable(&MotionSensor);

  /* 3. Route BOTH Data Ready signals to INT1 */
  ISM330DHCX_Set_INT1_Drdy(&MotionSensor, ENABLE); 
}
```

Update your while(1) loop logging:
To keep things clean for Python, we will log in this format: AccX,AccY,AccZ,GyroX,GyroY,GyroZ.


```while (1)
{
  if (dataRdyIntReceived > 0) 
  {
    dataRdyIntReceived = 0;
    
    ISM330DHCX_Axes_t acc, gyro;
    ISM330DHCX_ACC_GetAxes(&MotionSensor, &acc);
    ISM330DHCX_GYRO_GetAxes(&MotionSensor, &gyro);

    /* Log 6-axis data: Values are converted to G's and dps (Degrees Per Second) */
    UART_Printf_DMA("%.3f,%.3f,%.3f,%.3f,%.3f,%.3f\r\n", 
                   (float)acc.x/1000.0f, (float)acc.y/1000.0f, (float)acc.z/1000.0f,
                   (float)gyro.x/1000.0f, (float)gyro.y/1000.0f, (float)gyro.z/1000.0f);
  }
}```

2. Data Normalization (Pre-processing)

Neural networks are sensitive to "scale." Accelerometer values usually sit between −2 and +2 Gs, but Gyroscope values can spike to ±2000 dps. If you feed these raw numbers in, the model will "listen" to the Gyroscope more because the numbers are bigger.

Scaling Formula:
We scale both to a roughly similar range (typically [−1,1]):
Accscaled​=Accmg​​/1000
Gyroscaled​=  Gyromdps​​/1000×250

(Dividing gyro by 250 puts a 250 dps rotation at 1.0, keeping it in range with the accelerometer.)
